File: RNConekta/RNConekta/RNConekta.m
//
//  RNConekta.m

#import <UIKit/UIKit.h>
#import "Conekta.h"
#import "RNConekta.h"
#import <React/RCTConvert.h>
#import <React/RCTEventDispatcher.h>
#import <React/RCTView.h>
#import <React/UIView+React.h>
#import <React/RCTRootView.h>


@implementation RNConekta
    Conekta *conekta;
    BOOL isCollected = NO;

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

+ (BOOL)requiresMainQueueSetup
{
    return YES;
}

- (id)init
{
    UIViewController *controller = [[UIViewController alloc] init];

    conekta = [[Conekta alloc] init];
    [conekta setDelegate: controller];

    return self;
}

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(createToken:(NSDictionary *)info
                  callback:(RCTResponseSenderBlock)successCallback
                  errorCallback:(RCTResponseSenderBlock)failureCallback)
{

    NSString *publicKey = [RCTConvert NSString:info[@"publicKey"]];
    [conekta setPublicKey:publicKey];

    if ( isCollected == NO ) {
        isCollected = YES;
        [conekta collectDevice];
    }

    NSString *cardNumber = [RCTConvert NSString:info[@"cardNumber"]];
    NSString *cardHolder = [RCTConvert NSString:info[@"name"]];
    NSString *cvc = [RCTConvert NSString:info[@"cvc"]];
    NSString *expMonth = [RCTConvert NSString:info[@"expMonth"]];
    NSString *expYear = [RCTConvert NSString:info[@"expYear"]];
    
    Card *card = [conekta.Card initWithNumber: cardNumber name: cardHolder cvc: cvc expMonth: expMonth expYear: expYear];
    
    Token *token = [conekta.Token initWithCard:card];
    
    [token createWithSuccess: ^(NSDictionary *data) {
        successCallback(@[data]);
    } andError: ^(NSError *error) {
        failureCallback(@[[NSNull null]]);
    }];
}

@end

File: android/src/main/java/com/dieam/reactnativeconekta/modules/RNConekta.java
package com.dieam.reactnativeconekta.modules;

import android.app.Activity;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;

import org.json.JSONObject;

import io.conekta.conektasdk.Card;
import io.conekta.conektasdk.Conekta;
import io.conekta.conektasdk.Token;

public class RNConekta extends ReactContextBaseJavaModule {
    Activity mActivity;
    Boolean isCollected = false;

    public RNConekta(ReactApplicationContext reactContext, Activity activity) {
        super(reactContext);
        mActivity = activity;
    }

    @Override
    public String getName() {
        return "RNConekta";
    }

    @ReactMethod
    public void createToken(ReadableMap info, final Callback successCallback, final Callback failureCallback) {

        String publicKey = info.getString("publicKey");

        Conekta.setPublicKey(publicKey);

        if ( isCollected ) {
            isCollected = true;
            Conekta.collectDevice(mActivity);
        }

        String cardNumber = info.getString("cardNumber");
        String name = info.getString("name");
        String cvc = info.getString("cvc");
        String expMonth = info.getString("expMonth");
        String expYear = info.getString("expYear");

        Card card = new Card(name, cardNumber, cvc, expMonth, expYear);
        Token token = new Token(mActivity);

        token.onCreateTokenListener( new Token.CreateToken(){
            @Override
            public void onCreateTokenReady(JSONObject data) {
                try {
                    successCallback.invoke(data.toString());
                } catch (Exception err) {
                    failureCallback.invoke(err.getMessage());
                }
            }
        });

        token.create(card);
    }
}